### PR TITLE
[integration tests] Pre issue access token with auth code grant type

### PR DESCRIPTION
**Pull request by Shenali-SJ**
### Description
This adds integration tests for the pre issue access token action execution flow with the authorization code grant type.
This covers the test cases for the following scenarios.
- Add, replace, remove custom scopes
- Add, replace, remove `aud` claim
- Add custom claims - (data types supported: string, number, boolean, string array)
- Replace `expires_in` claim

A mock server has been set up using Wiremock to simulate the external service. The mock server runs locally on the port `8587`.

### Related Issue
- https://github.com/wso2/product-is/issues/20842

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java [hunk 1]
```diff
@@ -179,34 +212,57 @@ protected boolean isRedirectable(String method) {
...
+        requestedScopes.addAll(customScopes);
 
-        actionsMockServer.createMockServer(MOCK_SERVER_ENDPOINT);
         actionId = createPreIssueAccessTokenAction();
+
+        try {
+            Thread.sleep(3600);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
```
Comment (malithie) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java:240]: why do we have this ?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java [hunk 2]
```diff
@@ -322,7 +379,95 @@ public void testTokenExpiresInClaimReplaceOperation() throws Exception {
...
+        UserStore userStore =
+                new UserStore(Base64.getEncoder().encodeToString("PRIMARY".getBytes(StandardCharsets.UTF_8)),
+                        "PRIMARY");
+
+        PreIssueAccessTokenEvent event = new PreIssueAccessTokenEvent.Builder()
+                .request(tokenRequest)
+                .accessToken(accessTokenInRequest)
+                .tenant(tenant)
+                .organization(null)
+                .user(user)
```
Comment (malithie) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java:406]: For client credentials ideally there should not be a user 
Is there a user in the request ?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+                PROFILE_SCOPE);
+        requestedScopes.addAll(customScopes);
+
+        actionId = createPreIssueAccessTokenAction();
+
+        try {
+            Thread.sleep(3600);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
```
Comment (malithie) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java:251]: why do we have this

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+
+        ApplicationConfig applicationConfig = new ApplicationConfig.Builder()
+                .claimsList(userClaimConfigs)
+                .grantTypes(new ArrayList<>(Collections.singleton(OAUTH2_GRANT_TYPE_AUTHORIZATION_CODE)))
+                .tokenType(ApplicationConfig.TokenType.JWT)
+                .expiryTime(3600)
+                .skipConsent(true)
+                .build();
+
+        appTokenScopes = new TokenScopes.Builder().requestedScopes(Arrays.asList(OPENID_SCOPE, PROFILE_SCOPE,
```
Comment (ashanthamara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java:688]: Why do we have this inside this method ? And can't see any usages of this.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java [hunk 5]
```diff
@@ -179,34 +212,57 @@ protected boolean isRedirectable(String method) {
...
+
         if (!CarbonUtils.isLegacyAuthzRuntimeEnabled()) {
-            authorizeSystemAPIs(applicationId, new ArrayList<>(Arrays.asList(SCIM2_USERS_API, ACTIONS_API,
-                    APPLICATION_MANAGEMENT_API, API_RESOURCE_MANAGEMENT_API)));
+            authorizeSystemAPIs(applicationId, Collections.singletonList(SCIM2_USERS_API));
         }
         domainAPIId = createDomainAPI(EXTERNAL_SERVICE_NAME, EXTERNAL_SERVICE_URI, customScopes);
         authorizeDomainAPIs(applicationId, domainAPIId, customScopes);
-
         addUserWithRole(applicationId, customScopes);
```
Comment (Shenali-SJ) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java:223]: Remove this section as this is client credentials grant type.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java [hunk 6]
```diff
@@ -179,34 +212,57 @@ protected boolean isRedirectable(String method) {
...
 
-        actionsMockServer = new ActionsMockServer();
+        requestedScopes = new ArrayList<>();
+        Collections.addAll(requestedScopes,
+                INTERNAL_ORG_USER_MANAGEMENT_LIST,
```
Comment (Shenali-SJ) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java:227]: Update the scope

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java (hunk 1 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java]: This was removed with e641385fd8a50b37b7b3090ff41135d479266800

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java (hunk 3 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java]: This was removed with e641385fd8a50b37b7b3090ff41135d479266800

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java (hunk 4 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java]: This is refactored with e641385fd8a50b37b7b3090ff41135d479266800

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java (hunk 5 above)
Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenClientCredentialsGrantTestCase.java]: This is refactored with e641385fd8a50b37b7b3090ff41135d479266800

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/model/PreIssueAccessTokenEvent.java [hunk 7]
```diff
@@ -54,15 +54,21 @@ public AccessToken getAccessToken() {
...
-                Objects.equals(tenant, that.tenant) &&
-                Objects.equals(user, that.user) &&
-                Objects.equals(userStore, that.userStore);
+                Objects.equals(tenant, that.tenant);
+
+        if (!"client_credentials".equals(that.request.getGrantType())) {
+            return isEqualGeneral &&
+                    Objects.equals(user, that.user) &&
+                    Objects.equals(userStore, that.userStore);
+        }
```
Comment (malithie) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/model/PreIssueAccessTokenEvent.java:70]: Is this really necessary. If it's client credential user, user store will be null. And this should be verified at validation. Why we want to check for client credentials and behave differently

Comment (Shenali-SJ) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/model/PreIssueAccessTokenEvent.java:70]: When we are using the `AssertEquals` method of the `TestNG` framework within `testPreIssueAccessTokenActionRequest` test method, it internally invokes the equals method of each nested object. Therefore, during the comparison of `PreIssueAccessTokenActionRequest.PreIssueAccessTokenEvent `object in the `client credentials` grant flow the presence of user-related fields causes the assertion to fail as there is no user object associated. Hence, we have introduced a conditional check to ensure that the unnecessary fields are excluded from the equality determination only for this specific grant type.

As a further improvement, we could consider making User and UserStore Optional fields within the Event class.

**Review by malithie — APPROVED** [review]

**Pull request merged**